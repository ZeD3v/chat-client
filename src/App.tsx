import React, { useState, Dispatch, SetStateAction, useMemo } from 'react';

import Login from './pages/Login';
import Chat from './pages/Chat';

import styles from './common/styles/App.module.scss';

export type User = {
  username: string;
  token: string;
};

type UserContext = {
  user: User | null;
  setUser: Dispatch<SetStateAction<User | null>> | null;
};

type ErrorContext = {
  error: string | null;
  setError: Dispatch<SetStateAction<string | null>> | null;
};

export const UserContext = React.createContext<UserContext>({
  user: null,
  setUser: null,
});

export const ErrorContext = React.createContext<ErrorContext>({
  error: null,
  setError: null,
});

const App: React.FC = () => {
  const [user, setUser] = useState<User | null>(null);
  const [error, setError] = useState<string | null>(null);

  const userContextValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const errorContextValue = useMemo(() => ({ error, setError }), [
    error,
    setError,
  ]);

  return (
    <UserContext.Provider value={userContextValue}>
      <ErrorContext.Provider value={errorContextValue}>
        <div className={styles.container}>
          <div className={styles['inner-container']}>
            {/*
          For this simple scenario, basic if block will do.
          When page count grows and url navigation support becomes
          necessary - react-router or some similar navigation solution
          should be introduced.
        */}
            {(!user && <Login />) || <Chat />}
          </div>
        </div>
      </ErrorContext.Provider>
    </UserContext.Provider>
  );
};

export default App;
