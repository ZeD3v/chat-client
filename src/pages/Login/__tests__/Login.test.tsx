import React from 'react';
import fetchMock from 'jest-fetch-mock';
import { render, fireEvent, getByTestId, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { UserContext, ErrorContext } from '../../../App';
import Login from '../index';

describe('Login', () => {
  test('renders correctly', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('sends login request when login button is clicked', async () => {
    const { container } = render(
      <UserContext.Provider
        value={{
          user: { username: 'Test user', token: 'test-token' },
          setUser: jest.fn(),
        }}
      >
        <ErrorContext.Provider
          value={{
            error: null,
            setError: jest.fn(),
          }}
        >
          <Login />
        </ErrorContext.Provider>
      </UserContext.Provider>
    );
    const testUsername = 'tester';

    fireEvent.change(getByTestId(container, 'login-username-input'), {
      target: { value: testUsername },
    });

    // eslint-disable-next-line @typescript-eslint/require-await
    await act(async () => {
      fireEvent.click(getByTestId(container, 'login-submit-btn'));
    });

    expect(fetchMock).toHaveBeenCalled();
    expect(fetchMock).toHaveBeenCalledWith('http://localhost:3001/login', {
      body: `{"username":"${testUsername}"}`,
      headers: { 'Content-Type': 'application/json' },
      method: 'POST',
    });
  });

  test('show error message if error is set', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: { username: 'Test user', token: 'test-token' },
          setUser: jest.fn(),
        }}
      >
        <ErrorContext.Provider
          value={{
            error: 'Something went wrong',
            setError: jest.fn(),
          }}
        >
          <Login />
        </ErrorContext.Provider>
      </UserContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
