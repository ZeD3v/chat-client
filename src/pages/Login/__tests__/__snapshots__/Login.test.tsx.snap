// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="image-container"
    >
      <div
        class="image-inner-container"
      >
        <svg
          style="fill: #eee;"
          viewBox="0 -26 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 100c-5.52 0-10 4.48-10 10s4.48 10 10 10 10-4.48 10-10-4.48-10-10-10zm0 0M90 280c5.52 0 10-4.48 10-10s-4.48-10-10-10-10 4.48-10 10 4.48 10 10 10zm0 0"
          />
          <path
            d="M336 0c-90.027 0-163.918 62.07-169.633 140.254C80.63 144.554 0 206.379 0 290c0 34.945 13.828 68.805 39 95.633 4.98 20.531-1.066 42.293-16.07 57.297A9.998 9.998 0 0030 460c28.52 0 56.004-11.184 76.426-30.89C126.32 435.89 152.277 440 176 440c90.016 0 163.898-62.055 169.629-140.223 20.937-.93 42.715-4.797 59.945-10.668C425.996 308.816 453.48 320 482 320a9.998 9.998 0 007.07-17.07c-15.004-15.004-21.05-36.766-16.07-57.297 25.172-26.828 39-60.688 39-95.633C512 63.113 425.16 0 336 0zM176 420c-23.602 0-50.496-4.633-68.512-11.8a10 10 0 00-11.078 2.538c-12.074 13.2-27.773 22.403-44.879 26.633a80.872 80.872 0 006.098-59.52 9.98 9.98 0 00-2.445-4.226C32.496 350.258 20 320.559 20 290c0-70.469 71.438-130 156-130 79.852 0 150 55.527 150 130 0 71.684-67.29 130-150 130zm280.816-186.375a10.027 10.027 0 00-2.445 4.227 80.872 80.872 0 006.098 59.52c-17.106-4.227-32.805-13.435-44.88-26.634a10.007 10.007 0 00-11.077-2.539c-15.614 6.211-37.887 10.512-58.914 11.551-2.922-37.816-21.786-73.36-54.036-99.75H422c5.523 0 10-4.477 10-10s-4.477-10-10-10H260.84c-22.7-11.555-48.188-18.293-74.422-19.707C192.164 73.129 257.058 20 336 20c84.563 0 156 59.531 156 130 0 30.559-12.496 60.258-35.184 83.625zm0 0"
          />
          <path
            d="M256 260H130c-5.523 0-10 4.477-10 10s4.477 10 10 10h126c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0M256 320H90c-5.523 0-10 4.477-10 10s4.477 10 10 10h166c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0M422 100H296c-5.523 0-10 4.477-10 10s4.477 10 10 10h126c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0"
          />
        </svg>
      </div>
    </div>
    <div
      class="form-container"
    >
      <form
        class="form"
      >
        <input
          class="username-input"
          data-testid="login-username-input"
          minlength="3"
          name="username"
          placeholder="Username"
          required=""
          type="text"
          value=""
        />
        <button
          class="button submit-button"
          data-testid="login-submit-btn"
          type="submit"
        >
          Join chat
        </button>
      </form>
      <div
        class="container container--overlay container--hidden"
        style="background: rgb(236, 236, 236);"
      >
        <div
          class="inner-container"
        >
          <div
            class="sphere"
          />
          <div
            class="sphere sphere--2"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login show error message if error is set 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="image-container"
    >
      <div
        class="image-inner-container"
      >
        <svg
          style="fill: #eee;"
          viewBox="0 -26 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 100c-5.52 0-10 4.48-10 10s4.48 10 10 10 10-4.48 10-10-4.48-10-10-10zm0 0M90 280c5.52 0 10-4.48 10-10s-4.48-10-10-10-10 4.48-10 10 4.48 10 10 10zm0 0"
          />
          <path
            d="M336 0c-90.027 0-163.918 62.07-169.633 140.254C80.63 144.554 0 206.379 0 290c0 34.945 13.828 68.805 39 95.633 4.98 20.531-1.066 42.293-16.07 57.297A9.998 9.998 0 0030 460c28.52 0 56.004-11.184 76.426-30.89C126.32 435.89 152.277 440 176 440c90.016 0 163.898-62.055 169.629-140.223 20.937-.93 42.715-4.797 59.945-10.668C425.996 308.816 453.48 320 482 320a9.998 9.998 0 007.07-17.07c-15.004-15.004-21.05-36.766-16.07-57.297 25.172-26.828 39-60.688 39-95.633C512 63.113 425.16 0 336 0zM176 420c-23.602 0-50.496-4.633-68.512-11.8a10 10 0 00-11.078 2.538c-12.074 13.2-27.773 22.403-44.879 26.633a80.872 80.872 0 006.098-59.52 9.98 9.98 0 00-2.445-4.226C32.496 350.258 20 320.559 20 290c0-70.469 71.438-130 156-130 79.852 0 150 55.527 150 130 0 71.684-67.29 130-150 130zm280.816-186.375a10.027 10.027 0 00-2.445 4.227 80.872 80.872 0 006.098 59.52c-17.106-4.227-32.805-13.435-44.88-26.634a10.007 10.007 0 00-11.077-2.539c-15.614 6.211-37.887 10.512-58.914 11.551-2.922-37.816-21.786-73.36-54.036-99.75H422c5.523 0 10-4.477 10-10s-4.477-10-10-10H260.84c-22.7-11.555-48.188-18.293-74.422-19.707C192.164 73.129 257.058 20 336 20c84.563 0 156 59.531 156 130 0 30.559-12.496 60.258-35.184 83.625zm0 0"
          />
          <path
            d="M256 260H130c-5.523 0-10 4.477-10 10s4.477 10 10 10h126c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0M256 320H90c-5.523 0-10 4.477-10 10s4.477 10 10 10h166c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0M422 100H296c-5.523 0-10 4.477-10 10s4.477 10 10 10h126c5.523 0 10-4.477 10-10s-4.477-10-10-10zm0 0"
          />
        </svg>
      </div>
    </div>
    <div
      class="form-container"
    >
      <div
        class="error"
      >
        <div
          class="error-icon"
        >
          <svg
            style="width: 100%; height: 100%;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M501.609 384.603L320.543 51.265c-13.666-23.006-37.802-36.746-64.562-36.746-26.76 0-50.896 13.74-64.562 36.746-.103.176-.19.352-.293.528L10.662 384.076c-13.959 23.491-14.223 51.702-.719 75.457 13.535 23.769 37.919 37.948 65.266 37.948h360.544c27.347 0 52.733-14.179 66.267-37.948 13.504-23.754 13.241-51.967-.411-74.93zM225.951 167.148c0-16.586 13.445-30.03 30.03-30.03 16.586 0 30.03 13.445 30.03 30.03v120.121c0 16.584-13.445 30.03-30.03 30.03s-30.03-13.447-30.03-30.03V167.148zm30.03 270.273c-24.839 0-45.046-20.206-45.046-45.046 0-24.839 20.206-45.045 45.046-45.045 24.839 0 45.045 20.206 45.045 45.045.001 24.839-20.205 45.046-45.045 45.046z"
            />
          </svg>
        </div>
        <div>
          Something went wrong
        </div>
      </div>
      <form
        class="form"
      >
        <input
          class="username-input"
          data-testid="login-username-input"
          minlength="3"
          name="username"
          placeholder="Username"
          required=""
          type="text"
          value=""
        />
        <button
          class="button submit-button"
          data-testid="login-submit-btn"
          type="submit"
        >
          Join chat
        </button>
      </form>
      <div
        class="container container--overlay container--hidden"
        style="background: rgb(236, 236, 236);"
      >
        <div
          class="inner-container"
        >
          <div
            class="sphere"
          />
          <div
            class="sphere sphere--2"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
